src/library/scala/collection/immutable/ChampHashMap.scala

```diff
@@ -762,16 +829,299 @@ object HashMap extends MapFactory[HashMap] {
+  private def updateIfNotExists(mapNode: MapNode[K, V], key: K, value: V, originalHash: Int, keyHash: Int, shift: Int): Unit = {
+    mapNode match {
+      case bm: BitmapIndexedMapNode[K, V] =>
+        val mask = maskFrom(keyHash, shift)
+        val bitpos = bitposFrom(mask)
+        if ((bm.dataMap & bitpos) != 0) {
+          val index = indexFrom(bm.dataMap, mask, bitpos)
+          val key0 = bm.getKey(index)
+
+          if (key0 == key) {
+            () // do nothing
+          } else {
+            ensureUnaliased()
+
+            val value0 = bm.getValue(index)
+            val key0UnimprovedHash = bm.getHash(index)
+            val key0Hash = improve(key0UnimprovedHash)
+
+            val subNodeNew: MapNode[K, V] =
+              bm.mergeTwoKeyValPairs(key0, value0, key0UnimprovedHash, key0Hash, key, value, originalHash, keyHash, shift + BitPartitionSize)
+
+            hash += keyHash
+            migrateFromInlineToNode(bm, bitpos, subNodeNew)
+          }
+
+        } else if ((bm.nodeMap & bitpos) != 0) {
+          val index = indexFrom(bm.nodeMap, mask, bitpos)
+          val subNode = bm.getNode(index)
+          val beforeSize = subNode.size
+          updateIfNotExists(subNode, key, value, originalHash, keyHash, shift + BitPartitionSize)
+          bm.size += subNode.size - beforeSize
+        } else {
+          ensureUnaliased()
+          insertValue(bm, bitpos, key, originalHash, keyHash, value)
+          hash += keyHash
+        }
+      case hc: HashCollisionMapNode[K, V] =>
+        val index = hc.indexOf(key)
+        if (index < 0) {
+          ensureUnaliased()
+          hash += keyHash
+          hc.content = hc.content.appended((key, value))
+        }
+    }
+  }
+
+
+  /** If currently referencing aliased structure, copy elements to new mutable structure */
+  private def ensureUnaliased() = {
+    if (isAliased) copyElems()
+    aliased = null
+  }
+
+  /** Copy elements to new mutable structure */
+  private def copyElems(): Unit = {
+    aliased = null
+    rootNode = rootNode.copy()
+  }
+
+  override def result(): HashMap[K, V] =
+    if (rootNode.size == 0) {
+      HashMap.empty
+    } else if (aliased != null) {
+      aliased
+    } else {
+      aliased = new HashMap(rootNode, hash)
+      releaseFence()
+      aliased
+    }
+
+  override def addOne(elem: (K, V)): this.type = {
+    ensureUnaliased()
+    val h = elem._1.##
+    val im = improve(h)
+    update(rootNode, elem._1, elem._2, h, im, 0)
+    this
+  }
+
+
+  private[collection] def addOneIfNotExists(key: K, value: V): this.type = {
+    val h = key.##
+    val im = improve(h)
+    updateIfNotExists(rootNode, key, value, h, im, 0)
+    this
+  }
+
+  override def addAll(xs: IterableOnce[(K, V)]) = {
+    ensureUnaliased()
+    xs match {
+      case hm: HashMap[K, V] =>
+        def addAllRec(node: MapNode[K, V]): Unit = node match {
```

Comment:
Champ iterators are [implemented ](https://github.com/scala/scala/blob/88456afa67011681dc031533060df6f6d5b7b37c/src/library/scala/collection/immutable/ChampCommon.scala#L91-L99) in a fixed stack size algorithm by using data structure to track the cursor to the focussed node, rather than the JVM stack. This iteration doesn't follow that pattern. I would prefer we use a consistent approach.

Reply:
👍 
Although, I think this method can be further optimized by "zipping together" the two hashmaps, nod-by-node, rather than updating from the top node each time. That can be for an other PR though.

Reply:
Yes, the zipping approach sounds ideal.

Reply:
FYI, I've tried to improve the efficency of `ChampBaseIterator` for small collections in  #7224.